File: src/com/sterndu/data/transfer/basic/Socket.java
package com.sterndu.data.transfer.basic;

import java.io.*;
import java.net.*;
import java.nio.*;
import java.security.*;
import java.util.*;
import java.util.concurrent.CancellationException;
import java.util.function.*;
import com.sterndu.data.transfer.*;
import com.sterndu.multicore.Updater;
import com.sterndu.util.Util;
import com.sterndu.util.interfaces.ThrowingRunnable;

public class Socket extends DatatransferSocket {

	protected boolean host = false;

	protected Socket() {}

	public Socket(InetAddress address, int port) throws IOException {
		super(address, port);
		init(host);
	}

	public Socket(InetAddress address, int port, InetAddress localAddr, int localPort)
			throws IOException {
		super(address, port, localAddr, localPort);
		init(host);
	}

	public Socket(String host, int port) throws UnknownHostException, IOException {
		super(host, port);
		init(this.host);
	}

	public Socket(String host, int port, InetAddress localAddr, int localPort) throws IOException {
		super(host, port, localAddr, localPort);
		init(this.host);
	}

	protected byte[] implRecieveData(byte type, byte[] data) {
		return data;
	}

	protected byte[] implSendData(byte type, byte[] data) {
		return data;
	}

	protected void init(boolean host) {
		try {
			md = MessageDigest.getInstance("SHA3-256");
			setHandle((byte) -1, (type, data) -> {
				try {
					close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			});
		} catch (final NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		Updater.getInstance().add((ThrowingRunnable) () -> {
			if (!isClosed() && getInputStream().available() > 0) try {
				final Packet data = recieveData();
				if (handles.containsKey(data.type())) getHandle(data.type()).accept(data.type(), data.data());
				else recvVector.add(data);
			} catch (final CancellationException e) {

			}
			if (delayed_send.size() > 0 && initialized) {
				final Packet data = delayed_send.remove(0);
				sendData(data.type(), data.data());
			}
		}, "CheckForMsgs" + hashCode());
	}

	protected final Packet recieveData() throws IOException {
		Packet packet;
		sync: synchronized (recLock) {
			byte[] b = new byte[5];
			if (Util.readXBytes(b, getInputStream(), b.length)) {
				final byte type = b[0];
				final int length = ByteBuffer.wrap(b).order(ByteOrder.BIG_ENDIAN).getInt(1);
				b = new byte[32];
				final byte[] data = new byte[length];
				if (Util.readXBytes(data, getInputStream(), length)
						&& Util.readXBytes(b, getInputStream(), b.length) && Arrays.equals(b, md.digest(data))) {
					packet = new Packet(type, data);
					if (System.getProperty("debug").equals("true"))
						System.err.println(type + "r[length:" + length + ",data:" + Arrays.toString(data) + ",hash:"
								+ Arrays.toString(b));
					break sync;
				}
				if (System.getProperty("debug").equals("true"))
					System.out.println(type + "f[length:" + length + ",data:" + Arrays.toString(data) + ",hash:"
							+ Arrays.toString(b));
			}
			if (System.getProperty("debug").equals("true"))
				System.out.println("f" + Arrays.toString(b));
			return new Packet((byte) -128, new byte[0]);
		}
		return new Packet(packet.type(), implRecieveData(packet.type(), packet.data()));
		// type byte; length int; data byte[]; hash byte[];
	}

	protected void setHost(boolean host) { this.host = host; }

	@Override
	public void close() throws IOException, SocketException {
		try {
			synchronized (recLock) {
				synchronized (sendLock) {
					if (!isClosed()) {
						shutdownHook.accept(this);
						shutdownOutput();
						shutdownInput();
						Updater.getInstance().remove("CheckForMsgs" + hashCode());
						super.close();
					}
				}
			}
		} catch (NullPointerException e) {
			Updater.getInstance().remove("CheckForMsgs" + hashCode());
			super.close();
		}
	}

	public final BiConsumer<Byte, byte[]> getHandle(byte type) {
		return hasHandle(type) ? handles.get(type).getValue() : null;
	}

	public int getMessageCount() { return recvVector.size(); }

	public Packet getMessageFromBuffer() {
		if (recvVector.isEmpty()) throw new EmptyStackException();
		return recvVector.remove(0);
	}

	public final boolean hasHandle(byte type) {
		return handles.containsKey(type);
	}

	public boolean hasMessage() {
		return !recvVector.isEmpty();
	}

	public boolean isHost() { return host; }

	public void requestResend() throws SocketException {
		sendData((byte) 0, new byte[0]);
	}

	public void sendClose() throws SocketException {
		sendData((byte) -1, new byte[0]);
	}

	public final void sendData(byte type, byte[] data) throws SocketException {
		if (!initialized) try {
			final Class<?> caller = Class.forName(Thread.currentThread().getStackTrace()[2].getClassName());
			if (System.getProperty("debug").equals("true"))
				System.out.println(caller);
			caller.asSubclass(Socket.class);
		} catch (ClassNotFoundException | ClassCastException e) {
			delayed_send.add(new Packet(type, data));
			return;
		}
		if (!isClosed()) synchronized (sendLock) {
			final byte[] modified_data = implSendData(type, data);
			final byte[] hash = md.digest(modified_data);
			final byte[] length_bytes = ByteBuffer.allocate(4).putInt(modified_data.length).array();
			try {
				final OutputStream os = getOutputStream();
				os.write(type);
				os.write(length_bytes);
				os.write(modified_data);
				os.write(hash);
				if (System.getProperty("debug").equals("true"))
					System.err.println(
							type + "s[length_bytes:" + Arrays.toString(length_bytes) + ", length:"
									+ modified_data.length
									+ ",data:" + Arrays.toString(modified_data) + ",hash:" + Arrays.toString(hash));
			} catch (final IOException e) {
				e.printStackTrace();
				delayed_send.add(new Packet(type, data));
				return;
			}
		}
		else throw new SocketException("Soket closed!");

	}

	public final boolean setHandle(byte type, BiConsumer<Byte, byte[]> handle) {
		try {
			final Class<?> caller = Class.forName(Thread.currentThread().getStackTrace()[2].getClassName());
			if (!handles.containsKey(type)) {
				handles.put(type, Map.entry(caller, handle));
				return true;
			} else if (handles.get(type).getKey().equals(caller)) {
				handles.put(type, Map.entry(caller, handle));
				return true;
			} else return false;
		} catch (final Exception e) {
			e.printStackTrace();
			return false;
		}
	}

	public void setShutdownHook(Consumer<DatatransferSocket> hook) { shutdownHook = hook; }

}

File: src/com/sterndu/data/transfer/basic/Socket.kt
@file:JvmName("Socket")
package com.sterndu.data.transfer.basic

import com.sterndu.data.transfer.DataTransferClient
import com.sterndu.data.transfer.Packet
import com.sterndu.multicore.LoggingUtil
import com.sterndu.multicore.Updater
import com.sterndu.util.readXBytes
import java.io.File
import java.io.IOException
import java.net.InetAddress
import java.net.SocketException
import java.net.UnknownHostException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger
import java.net.Socket as NetSocket

open class Socket(val socket: NetSocket = NetSocket()) : DataTransferClient() {

	private var logger: Logger = LoggingUtil.getLogger(basicSocket)

	override var appendix: String = "uninitialized socket"

	init {
		allSockets[this] = Thread.currentThread().stackTrace.let { it.copyOfRange(1, 4.coerceAtMost(it.size)) }
		if (socket.isBound) {
			appendix = "${socket.inetAddress}:${socket.port} -- ${socket.localAddress}:${socket.localPort}".replace("/", "").replace(":", "-")
			init(false)
		}
	}

	/**
	 * Instantiates a new socket.
	 *
	 * @param address the address
	 * @param port the port
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	@Throws(IOException::class)
	@Deprecated("Use Socket(socket: NetSocket) instead", ReplaceWith("Socket(NetSocket(address, port))", "import java.net.Socket as NetSocket"))
	constructor(address: InetAddress, port: Int) : this(NetSocket(address, port))

	/**
	 * Instantiates a new socket.
	 *
	 * @param address the address
	 * @param port the port
	 * @param localAddr the local addr
	 * @param localPort the local port
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	@Throws(IOException::class)
	@Deprecated("Use Socket(socket: NetSocket) instead", ReplaceWith("Socket(NetSocket(address, port, localAddr, localPort))", "import java.net.Socket as NetSocket"))
	constructor(address: InetAddress, port: Int, localAddr: InetAddress, localPort: Int) : this(NetSocket(
		address,
		port,
		localAddr,
		localPort
	))

	/**
	 * Instantiates a new socket.
	 *
	 * @param host the host
	 * @param port the port
	 * @throws UnknownHostException the unknown host exception
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	@Throws(IOException::class, UnknownHostException::class)
	@Deprecated("Use Socket(socket: NetSocket) instead", ReplaceWith("Socket(NetSocket(host, port))", "import java.net.Socket as NetSocket"))
	constructor(host: String, port: Int) : this(NetSocket(host, port))

	/**
	 * Instantiates a new socket.
	 *
	 * @param host the host
	 * @param port the port
	 * @param localAddr the local addr
	 * @param localPort the local port
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	@Throws(IOException::class)
	@Deprecated("Use Socket(socket: NetSocket) instead", ReplaceWith("Socket(NetSocket(host, port, localAddr, localPort))", "import java.net.Socket as NetSocket"))
	constructor(host: String, port: Int, localAddr: InetAddress, localPort: Int) : this(NetSocket(
		host,
		port,
		localAddr,
		localPort
	))

	fun initWithHost(host: Boolean) {
		appendix = "${socket.inetAddress}:${socket.port} -- ${socket.localAddress}:${socket.localPort}".replace("/", "").replace(":", "-")
		super.init(host)
	}

	override val isClosed: Boolean
    	get() = socket.isClosed

	override val isDataAvailable: Boolean
		get() = socket.inputStream.available() > 0

	override val isConnected: Boolean
		get() = socket.isConnected

	override fun name(withClassName: Boolean) = if (withClassName) "${javaClass.simpleName} ${socket.inetAddress}:${socket.port}" else "${socket.inetAddress}:${socket.port}"

	/**
	 * Receive data.
	 *
	 * @return the packet
	 * @throws IOException Signals that an I/O exception has occurred.
	 */
	@Throws(IOException::class)
    override fun receiveData(): Packet {
        val md = md ?: throw IllegalStateException()

        var packet: Packet
		var recvStamp = 0L
		try {
			recvStamp = recvLock.writeLock() // Get an exclusive lock for receiving

			var b = ByteArray(5)
			if (readXBytes(b, socket.inputStream, b.size, 5000)) {
				val type = b[0]
				val length = ByteBuffer.wrap(b).order(ByteOrder.BIG_ENDIAN).getInt(1)
                if (length in 0..MAX_PACKET_SIZE) {
                    b = ByteArray(32)
                    var data = ByteArray(length)
                    if (readXBytes(data, socket.inputStream, length, 5000 + length * 10L)
                        && readXBytes(b, socket.inputStream, b.size, 5000 + 320) && b.contentEquals(md.digest(data))
                    ) {
                        packet = Packet(type, data)
                        if (data.size > 5000) data = data.copyOfRange(0, 5000)
                        logReceiveState(type, 'r', length, data, b)
                    } else {
                        if (data.size > 5000) data = data.copyOfRange(0, 5000)
                        logReceiveState(type, 'f', length, data, b)
                        return Packet(((-128).toByte()), ByteArray(0))
                    }
                } else {
					logger.fine("$type f[length: $length]")
					return Packet(((-128).toByte()), ByteArray(0))
				}
			} else {
				logger.fine("f" + b.contentToString())
				return Packet(((-128).toByte()), ByteArray(0))
			}
		} finally {
			recvLock.unlock(recvStamp)
		}
		return Packet(packet.type, implReceiveData(packet.type, packet.data))
		// type byte; length int; data byte[]; hash byte[32];
	}

	private fun logReceiveState(type: Byte, state: Char, length: Int, data: ByteArray, hash: ByteArray) {
		if (state == 'r')
			logger.finest("$type $state[length: $length,data: ${data.contentToString()},hash: ${hash.contentToString()}")
		else
			logger.fine("$type $state[length: $length,data: ${data.contentToString()},hash: ${hash.contentToString()}")
	}

	/**
	 * Send data.
	 *
	 * @param type the type
	 * @param data the data
	 * @param raw true, if raw data should be sent
	 * @throws SocketException the socket exception
	 */
	@Throws(SocketException::class)
	override fun sendData(type: Byte, data: ByteArray, raw: Boolean) {
		if (!raw && !initialized) {
			delayedSend.add(Packet(type, data))
			return
		}
		if (isClosed) throw SocketException(socketClosed)
		val md = md ?: throw IllegalStateException()
		var sendStamp = 0L
		try {
			sendStamp = sendLock.writeLock() // Get an exclusive lock for sending

			if (isClosed) throw SocketException(socketClosed)
			Files.write(File("./${appendix}_${System.currentTimeMillis()}_${type}${if (raw) "I" else ""}S.pckt").toPath(), data, StandardOpenOption.CREATE, StandardOpenOption.WRITE) //write content -> appendix_timestamp.pckt
			val modifiedData = if (raw) data else implSendData(type, data)
			val hash = md.digest(modifiedData)
			val lengthBytes = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(modifiedData.size).array()
			try {
				val os = socket.outputStream
				os.write(type.toInt())
				os.write(lengthBytes)
				os.write(modifiedData)
				os.write(hash)
				logger.finest(type.toString() + "${if (raw) "i" else ""}s[length_bytes:" + lengthBytes.contentToString() + ", length:" + modifiedData.size +
						",data:" + modifiedData.contentToString() + ",hash:" + hash.contentToString()
				)
			} catch (e: SocketException) {
				try {
					close()
				} catch (ex: IOException) {
					ex.initCause(e)
					logger.log(Level.WARNING, basicSocket, ex)
				}
				if (e.message != "Broken pipe")
					throw e
				else logger.log(Level.FINEST, basicSocket, e)
			} catch (e: IOException) {
				logger.log(Level.WARNING, basicSocket, e)
				delayedSend.add(Packet(type, data))
			}
		} finally {
			sendLock.unlock(sendStamp)
		}
	}

	/**
	 * Close.
	 *
	 * @throws IOException Signals that an I/O exception has occurred.
	 * @throws SocketException the socket exception
	 */
	@Throws(IOException::class, SocketException::class)
	override fun close() {
		if (logger == null)
			logger = LoggingUtil.getLogger(basicSocket)
		logger.fine("close $this ${Thread.currentThread().stackTrace.contentToString()}")
		try {
			var recvStamp = 0L
			var sendStamp = 0L

			try {
				recvStamp = recvLock.writeLock() // Get an exclusive lock for receiving
				sendStamp = sendLock.writeLock() // Get an exclusive lock for sending

				try {
					if (!isClosed) {
						shutdownHook(this)
						socket.shutdownOutput()
						socket.shutdownInput()
						Updater.remove("CheckForMsgs $appendix")
						Updater.remove("PingKill $appendix")
						disablePeriodicPing()
						socket.close()
					}
				} catch (e: SocketException) {
					socket.close()
				}
			} finally {
				recvLock.unlock(recvStamp)
				sendLock.unlock(sendStamp)
			}
		} catch (_: NullPointerException) {
            Updater.remove("CheckForMsgs $appendix")
            Updater.remove("PingKill $appendix")
            disablePeriodicPing()
			socket.close()
		}
	}

	companion object {

		private const val MAX_PACKET_AMOUNT_FOR_RESEND = 10

		private const val basicSocket = "Basic Socket"

		private const val socketClosed = "Socket closed!"

		private const val MAX_PACKET_SIZE = 1_073_741_824 // 2^30

		val allSockets: MutableMap<Socket, Array<StackTraceElement>> = Collections.synchronizedMap(HashMap())
	}

}
